python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java

```diff
@@ -235,9 +243,32 @@ private static DeclaredType declaredTypeFromTypeAnnotation(Expression expression
     if (expression.is(Kind.NONE)) {
       return new DeclaredType(builtinSymbols.get(BuiltinTypes.NONE_TYPE));
     }
+    if (expression.is(Kind.BITWISE_OR)) {
+      var binaryExpression = (BinaryExpression) expression;
+      var unionType = declaredUnionType(binaryExpression.leftOperand(), binaryExpression.rightOperand(), builtinSymbols);
+      if (unionType != null) {
+        return unionType;
+      }
```

Comment:
You can probably just return unconditionally here, since the default is to return null.